Alex and Happy Part 2/article.md

```diff
@@ -0,0 +1,1323 @@
+
+A **non-terminal** is an arbitrary identifier that's replaced by other terminals and non-terminals (terminals and non-terminals together are called **symbols**) to produce derivations.
+In the example below, `myNonTerminal` is non-terminal.
+
+A **production rule** is a rule describing how to "produce" a string that is a sentence corresponding to the grammar.
+It has form `a : b { action }`, which looks like so in Happy:
+
+```happy
+myNonTerminal
+  : bodyA1 bodyA2 ... { actionA }
+  | bodyB1 bodyB2 ... { actionB }
+  | ...
+```
+
+n.b.: The ellipses here are not correct Happy grammar, and we use them to represent that more symbols may be present.
+
+`body@#` where `@` is some letter and `#` some number means that the parser must parse each body in succession.
+The pipe (`|`) represents an alternation. The parser may parse any of the given bodies, accepting the one that matches.
+An alternative may be empty, meaning that `myNonTerminal : { action }` is also accepted.
+`action@` represents what the parser should do once successfully parsed that rule. Like in Alex, it may be any Haskell expression. More about this later.
+
+Production rules may also have an optional type signature.
+For instance, the declaration of `myNonTerminal` could also have been written as such:
+
+```happy
+myNonTerminal :: { MyType }
+  : bodyA1 bodyA2 ... { actionA }
+  | bodyB1 bodyB2 ... { actionB }
+  | ...
+```
+
+Where any Haskell type may replace `MyType`.
+
+Later in the article, we will also use the following notation representing the same as above.
+This is not valid Happy code, but it's what it uses for debugging information.
+
+```
+myNonTerminal -> bodyA1 bodyA2 ...
+myNonTerminal -> bodyB1 bodyB2 ...
+myNonTerminal -> ...
+```
+
+There may be multiple productions in a single Happy file.
+
+### Parser position
+
+It's essential to introduce a notation that will be used later in this article.
+We use a dot (`.`) to represent the current position of the parser.
+For example, suppose we have the following:
+
+```
+exp -> exp . '+' exp
+```
+
+In this case, the parser has finished consuming the first `exp`, and it's about to consume the `+`.
+
+### Look-ahead
+
+The **look-ahead** represents the sequence of symbols that will be consumed next by the parser.
+Happy is an LALR(1) parser, which means its look-ahead is 1, so it can see one incoming symbol.
+In the `exp` example above, the look-ahead token will be `+`.
+
+### Actions (shift, reduce, goto, accept)
+
+When working with Happy, two core concepts that you need to know are **shifting** and **reducing**, which I'll briefly explain.
+
+An LR parser also contains a state machine, similarly to Alex.
+This finite-state machine is called an _LR automaton_, which is also a DFA, which is used to make parsing decisions.
+At the same time, the parser contains a stack with all the tokens consumed during parsing, making it a _pushdown automaton_.
```

Comment:
Is a finite state machine, a stack, and an action table all that is necessary to make a parser? 

Perhaps we can have a nice illustration here so that it is easily graspable what are the "objects" we are talking about.

Reply:
I think the user can see the demo below since it illustrates what is said here. At least, I'm not sure what would be the best illustration here otherwise...

Reply:
Traditionally (all coming back to the Dragon Book and probably even earlier), this kind of diagram is shown to compiler theory students with relation to LR parsing:

![image](https://user-images.githubusercontent.com/7275622/172454696-5cdfd6ad-308b-43f9-891f-17e3b0357acd.png)

I'm not sure if it's particularly useful however, in my experience those who understand this diagram, don't need it.